hta/trace_analysis.py

```diff
@@ -504,3 +505,61 @@ def get_cupti_counter_data_with_operators(
                 for individual performance counters.
         """
         return CuptiCounterAnalysis.get_counter_data_with_operators(self.t, ranks)
+
+    def critical_path_analysis_for_rank(
+        self,
+        rank: int,
+        annotation: str,
+        instance_id: Optional[int],
+    ) -> List[pd.DataFrame]:
+        r"""
+        Perform critical path analysis for trace events within a rank.
+        We further reduce the region of interest by selecting
+        a trace annotation and instance id. This will
+        limit the analysis to events within the time range of that annoation.
+        For example, you can use this to limit the analysis to one iteration
+        by passing annotation='ProfilerStep500'.
+
+        Args:
+            t (Trace): Input trace data structure.
+            rank (int): rank to analyze for the critical path.
+            annotation (str): a trace annotation to limit the analysis to,
+                such as "ProfilerStep1200"
+            instance_id (int): optionally specify which instance of the annotation
+                to consider. Defaults to the first instance.
+
+        Returns:
+            CPGraph object that can be used to obtain statistics and further
+            visualize the critical path.
+
+            CPGraph is also a subinstance of a networkx.DiGraph.
+            Run 'CPGraph?' for more info and APIs.
+        """
+        return CriticalPathAnalysis.critical_path_analysis_for_rank(
+            self.t, rank, annotation, instance_id
+        )
+
+    def overlay_critical_path_analysis_for_rank(
+        self,
+        rank: int,
+        critical_path_graph: CPGraph,
+        output_dir: str,
+        show_all_edges: bool = False,
+    ) -> str:
+        r"""
+        Overlay the identified critical path on top of the trace file
+        for visualization.
+
+        Args:
+            rank (int): rank to generate the time series for.
+            critical_path_graph: Critical Path Graph object generated previously
+            output_dir (str): Output director to store overlaid trace.
+            show_all_edges (bool): When set this will add edge events for
+                all types of edges in the critical path graph. This is useful
+                for debugging the algorithm.
+
+        Returns: the overlaid_trace_file path.
```

Comment:
please ensure there is no name conflict of the overlaid file with frequent cuda kernel patterns feature.